File: C++/1912 - Help Seu Madruga.cpp
#include <bits/stdc++.h>
using namespace std;


double cArea(double corte, vector <int>& v){
    double a = 0;
    for (double i:v){
        if (i>corte) a+= (i-corte);
    }
    return a;
}

void BB(int alvo, double e, double d, double meio, vector<int>& v){
    double area = cArea(meio, v);

    while (abs(area-alvo)>0.00001)    
    {
        if (area < alvo)
            d = meio;
        else{
            if (area > alvo)
                e = meio;
        }
        meio = (e+d)/2;
        area = cArea(meio,v);
    }

    cout<<fixed<<setprecision(4)<<meio<<endl;
    
}

void solve(int n, int a){

    vector <int> v(n);
    long soma = 0;
    int mai = 0;
    for (int i=0;i<n;i++){
        cin>>v[i];
        soma+=v[i];
        if (mai<v[i]) mai = v[i];
    }
    if (soma<a) cout<<"-.-"<<endl;
    else{
        if (soma==a) cout<<":D\n";
        else{
            double esq = 0;
            double dir = mai;
            double meio = (esq+dir)/2;
            BB(a,esq,dir,meio,v);
        }
    }

    
}

int main(){
    ios::sync_with_stdio(0);
    cin.tie(0);

    int n,a;
    
    while (true)
    {
        cin>>n>>a;
        if (n+a == 0)break;

        solve(n,a);
    }
    
}

File: PYTHON/1912 - Help Seu Madruga.py
import sys
ler = sys.stdin.readline

def cArea(corte):
    a = 0
    
    for i in range(int(corte),len(vf)):
        if vf[i]>0 and i>corte:
            a += (i-corte)*vf[i]
            
    return a


def BB(e,d,meio):
    area = cArea(meio)

    while abs(area-a) > 0.00001:
        #print(area,meio)
        
        if area < a:
            d = meio
        
        elif area > a:
            e = meio
        meio = (e+d)/2
        area = cArea(meio)
    print(f"{meio:.4f}")
         



while True:
    n,a = map(int,ler().split())
    if n+a ==0:
        break
    
    v = list(map(int,ler().split()))
    mai = max(v)
    vf = [0]*(mai+1)
    
    for i in v:
        vf[i]+=1
    #print(vf)
    
    soma = sum(v)
    if soma == a:
        print(":D")
    elif soma<a:
        print("-.-")
        
    else:
        esq = 0
        dir = max(v)
        meio = (esq+dir)/2
        
        BB(esq,dir,meio)
